rule(Change thread namespace): Modify condition to detect suspicious container activity

rung opened the pull request:
Signed-off-by: Hiroki Suezawa <suezawa@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. . Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**
/kind rule-update

**Any specific area of the project related to this PR?**
/area rules

**What this PR does / why we need it**:
- What this PR does
  - Modify condition to detect some process names within a container.
  - Add evt.dir to avoid detection of same syscall
  
- Why we need it
  - The current rule exclude some process names used on host
    - i.e. `k8s_binaries`, `nsenter`
  - But `nsenter` could be used to do privilege escalation
    - https://twitter.com/mauilion/status/1129468485480751104
    - (nsenter is included in many container images like `alpine` by default)
  - So I would like to change condition to detect suspicious container activity.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:


**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
rule(Change thread namespace): Modify condition to detect suspicious container activity
```

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **mstemm**

If they are not already assigned, you can assign the PR to them by writing `/assign @mstemm` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[rules/OWNERS](https://github.com/falcosecurity/falco/blob/dev/rules/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

rung commented:
### Procedure Step
- Start a container
```bash
kubectl run malicious-container --restart=Never -ti --rm --image lol --overrides '{"spec":{"hostPID": true, "containers":[{"name":"1","image":"alpine","command":["sh"],"stdin": true,"tty":true,"securityContext":{"privileged":true}}]}}'
```

- Change namespace to host 
```bash
nsenter --mount=/proc/1/ns/mnt -- /bin/bash
```

### Sample Output
```
{"output":"16:46:42.320531675: Notice Namespace change (setns) by unexpected program (user=root command=nsenter --mount=/proc/1/ns/mnt -- /bin/bash parent=sh k8s.ns=default k8s.pod=malicious-container container=3172964a195b container_id=3172964a195b image=alpine:latest) k8s.ns=default k8s.pod=malicious-container container=3172964a195b k8s.ns=default k8s.pod=malicious-container container=3172964a195b","priority":"Notice","rule":"Change thread namespace","time":"2019-12-16T16:46:42.320531675Z", "output_fields": {"container.id":"3172964a195b","container.image.repository":"alpine","container.image.tag":"latest","evt.time":1576514802320531675,"k8s.ns.name":"default","k8s.pod.name":"malicious-container","proc.cmdline":"nsenter --mount=/proc/1/ns/mnt -- /bin/bash","proc.pname":"sh","user.name":"root"}}
```

Kaizhe commented on rules/falco_rules.yaml:
```diff
@@ -1527,9 +1527,9 @@
-    evt.type = setns
-    and not proc.name in (docker_binaries, k8s_binaries, lxd_binaries, sysdigcloud_binaries,
-                          sysdig, nsenter, calico, oci-umount, network_plugin_binaries)
+    evt.type=setns and evt.dir=<
```
I like this change.`evt.dir=<`

Kaizhe commented:
/lgtm

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/dev/rules/OWNERS)~~ [Kaizhe]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: c19ce5e70645012684c11623453c3f717782b2ef</details>

rung commented:
I'll check why CI failed.

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*, *<a href="https://github.com/falcosecurity/falco/pull/974#pullrequestreview-333673615" title="Approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/dev/rules/OWNERS)~~ [Kaizhe,leodido]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

rung commented:
(hmm, I tried finding the reason to get error, but I haven't solved it yet)

Kaizhe commented:
@mstemm can you please help? The rule changes might break the existing tests. It would be great that you can point us to the right place. Thanks!

fntlnz commented:
Closing and reopening this to let this test against the new CI

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*, *<a href="https://github.com/falcosecurity/falco/pull/974#pullrequestreview-333673615" title="Approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [Kaizhe,leodido]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
@rung: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

rung commented:
oh, very sorry. I thought this was merged 🙇🙇.
I'll fix the confliction.

poiana commented:
@rung: Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository.
Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [Kaizhe]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [Kaizhe]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

rung commented on rules/falco_rules.yaml:
```diff
@@ -1544,9 +1544,9 @@
-                          sysdig, nsenter, calico, oci-umount, cilium-cni, network_plugin_binaries)
+    evt.type=setns and evt.dir=<
+    and not (container.id=host and proc.name in (docker_binaries, k8s_binaries, lxd_binaries, nsenter))
+    and not proc.name in (sysdigcloud_binaries, sysdig, calico, oci-umount, cilium-cni, network_plugin_binaries)
```
This line was conflicted because of the PR (https://github.com/falcosecurity/falco/commit/48a0f512fb9e5022135ca42040242bef19a98aea), and I fixed the conflict.

Kaizhe replied:
did you rebase?

rung replied:
Yes, after I got needs-rebase label from poiana. The point I changed in first commit is to add cilium-cni.

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **leodido**
You can assign the PR to them by writing `/assign @leodido` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["leodido"]} -->

rung commented:
### Report on deep dive (why regression tests failed).
#### Reason 1
- Related test cases
  - https://github.com/falcosecurity/falco/blob/master/test/falco_traces.yaml.in#L29
  - https://github.com/falcosecurity/falco/blob/master/test/falco_tests.yaml#L692
- How to fix
  - When I modified `"Change thread namespace": 2` to `"Change thread namespace": 1` on the above files, the test was successful.

- Reason
![image](https://user-images.githubusercontent.com/1150301/75301672-c0ae6680-587e-11ea-9970-712d3d69b766.png)
  - The scap file of regression test has 2 setns(enter and exit). but I added condition of `evt.dir=<`, so the count of detection should be 1 now.
  - Related files
    - Regression test script
      - https://github.com/falcosecurity/falco/blob/master/test/run_regression_tests.sh#L36
    - scap file for regression test
       - https://s3.amazonaws.com/download.draios.com/falco-tests/traces-positive.zip

#### Reason 2
- Error Message
```
15:30:21 ERROR| FAIL 55-/source/falco/test/falco_test.py:FalcoTest.test;no-kube-demo-47e1 -> TestFail: Detected 188 events when should have detected none
```

- `traces-negative/kube-demo.scap` has many `nsenter` processes
  - old rule bypassed the setns, but my original rule detected every `nsenter`. so the regression test failed.
  - it seems hyperkube(and kubelet) use `nsenter` binary

- Sample
```
DEBUG| [stdout] {"output":"04:46:52.887250174: Notice Namespace change (setns) by unexpected program (user=root command=nsenter --mount=/rootfs/proc/1/ns/mnt -- /bin/findmnt -o target --noheadings --target /var/lib/kubelet/pods/352aff16-2f8d-11e6-8bd9-0800270a6574/volumes/kubernetes.io~secret/default-token-pb8gr parent=hyperkube k8s-kubelet (id=af911d183773) container_id=af911d183773 image=<NA>:<NA>)","priority":"Notice","rule":"Change thread namespace","time":"2016-06-11T04:46:52.887250174Z", "output_fields": {"container.id":"af911d183773","container.image.repository":null,"container.image.tag":null,"container.name":"k8s-kubelet","evt.time":1465620412887250174,"proc.cmdline":"nsenter --mount=/rootfs/proc/1/ns/mnt -- /bin/findmnt -o target --noheadings --target /var/lib/kubelet/pods/352aff16-2f8d-11e6-8bd9-0800270a6574/volumes/kubernetes.io~secret/default-token-pb8gr","proc.pname":"hyperkube","user.name":"root"}}
```

- How to fix
  - add this line
```
not proc.pname in (hyperkube, kubelet)
```

rung commented:
@leodido @Kaizhe Could you review this PR when you have time? 
CI passed finally.
Thanks.

rung commented:
@leodido @Kaizhe How about this PR?
Thank you! 👍

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#pullrequestreview-373575908" title="Approved">fntlnz</a>*, *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: 17f1406746d56ff07ee0721850368ffb50565b69</details>

fntlnz commented:
Good job as always @rung !

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/974#pullrequestreview-373575908" title="Approved">fntlnz</a>*, *<a href="https://github.com/falcosecurity/falco/pull/974#issuecomment-566211129" title="LGTM">Kaizhe</a>*, *<a href="https://github.com/falcosecurity/falco/pull/974#pullrequestreview-373666050" title="Approved">leodido</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falco/blob/master/OWNERS)~~ [fntlnz,leodido]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

leodido commented:
/milestone 0.21.0

rung commented:
@fntlnz @leodido Thank you for your review!
